### MPG Dataset

In [66]:
import numpy as np
import pandas as pd

In [67]:
mpgDataSet=pd.read_csv("./mpg.csv")
mpg=pd.read_csv("./mpg.csv")

Q1. Inspect the data

Before touching anything, find out what you are holding.

1. How many rows and columns?
2. What is the dtype of each column?
3. Which columns have missing values, and how many?

In [68]:
row=mpg.shape[0]
col=mpg.shape[1]

print("Rows:",row,"Cols:",col)

Rows: 398 Cols: 9


In [69]:
print("Dtypes of each Column: \n", mpg.dtypes)

Dtypes of each Column: 
 mpg             float64
cylinders         int64
displacement    float64
horsepower      float64
weight            int64
acceleration    float64
model_year        int64
origin              str
name                str
dtype: object


In [70]:
print("Missing data in Column Details: \n",mpg.isnull().sum())

Missing data in Column Details: 
 mpg             0
cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
name            0
dtype: int64


Q2. Select columns

Show only name , mpg and origin , for the first 5 rows

In [71]:
showData=mpg[['name','mpg','origin']].head()
print("Data in first 5 rows:\n",showData)

Data in first 5 rows:
                         name   mpg origin
0  chevrolet chevelle malibu  18.0    usa
1          buick skylark 320  15.0    usa
2         plymouth satellite  18.0    usa
3              amc rebel sst  16.0    usa
4                ford torino  17.0    usa


Q3. Filter rows

How many cars manage more than 30 mpg? Of those, how many are
Japanese?

In [72]:
carsGreater30mpg=(mpg['mpg']>30).sum()
print("Cars with mpg greater than 30:",carsGreater30mpg)

japaneseCarGreater30mpg=((mpg['mpg']>30) & (mpg['origin']=='japan')).sum()
print("Japanese Cars with mpg greater than 30:",japaneseCarGreater30mpg)

Cars with mpg greater than 30: 85
Japanese Cars with mpg greater than 30: 46


Q4. Handle missing values
horsepower has blanks.

1. Which origins do the missing rows belong to?
2. Fill the blanks with the median horsepower.
3. Confirm there are no blanks left.

In [73]:
originsWithMissingHorsepower= mpg.loc[mpg['horsepower'].isnull(), 'origin']
print("Origin with Missing Horsepower:\n",originsWithMissingHorsepower)

Origin with Missing Horsepower:
 32        usa
126       usa
330    europe
336       usa
354    europe
374       usa
Name: origin, dtype: str


In [74]:
fillHorsepower=mpg['horsepower'].fillna((mpg['horsepower'].median()).round(2))
mpg['horsepower']=fillHorsepower
print("Filled Horsepower:\n",fillHorsepower)

Filled Horsepower:
 0      130.0
1      165.0
2      150.0
3      150.0
4      140.0
       ...  
393     86.0
394     52.0
395     84.0
396     79.0
397     82.0
Name: horsepower, Length: 398, dtype: float64


In [75]:
countNull=(mpg.isnull().sum()).sum()
print("Number of Null Data:",countNull)

Number of Null Data: 0


Q5. Create a new column

Power-to-weight ratio is a better predictor of how a car feels than horsepower
alone.
Add a column power_to_weight = horsepower ÷ weight. Then show the 5
cars with the highest ratio.

In [76]:
mpg['power_to_weight']=(mpg['horsepower']/mpg['weight']).round(2)
carsWithHighestRatio=mpg.sort_values(by='power_to_weight', ascending=False)
print("The first 5 cars with highest ratio:\n",carsWithHighestRatio.head())

The first 5 cars with highest ratio:
       mpg  cylinders  displacement  horsepower  weight  acceleration  \
13   14.0          8         455.0       225.0    3086          10.0   
116  16.0          8         400.0       230.0    4278           9.5   
96   13.0          8         360.0       175.0    3821          11.0   
23   26.0          4         121.0       113.0    2234          12.5   
27   11.0          8         318.0       210.0    4382          13.5   

     model_year  origin                     name  power_to_weight  
13           70     usa  buick estate wagon (sw)             0.07  
116          73     usa       pontiac grand prix             0.05  
96           73     usa  amc ambassador brougham             0.05  
23           70  europe                 bmw 2002             0.05  
27           70     usa               dodge d200             0.05  


Q6. Count with value_counts

1. How many cars come from each origin?
2. How many cars have each cylinder count?
3. Show the origin breakdown as percentages instead of counts.

In [77]:
carsFromEachOrigin=mpg['origin'].value_counts()
print("Cars from each origin:\n",carsFromEachOrigin)

Cars from each origin:
 origin
usa       249
japan      79
europe     70
Name: count, dtype: int64


In [78]:
carsCylinderCount=mpg['cylinders'].value_counts()
print("Cars with different cylinder values:\n",carsCylinderCount)

Cars with different cylinder values:
 cylinders
4    204
8    103
6     84
3      4
5      3
Name: count, dtype: int64


In [79]:
carsOriginPercent=mpg['origin'].value_counts(normalize=True)
print("Cars with origin as percent:",carsOriginPercent)

Cars with origin as percent: origin
usa       0.625628
japan     0.198492
europe    0.175879
Name: proportion, dtype: float64


Q7. Named aggregation

Build one table with a row per origin and these columns: cars , avg_mpg ,
avg_weight , avg_cylinders . Round to 1 decimal.

In [80]:
carDetails=mpg.groupby("origin").agg(
    cars=("name","size"),
    avg_mpg=("mpg","mean"),
    avg_weight=("weight","mean"),
    avg_cylinders=("cylinders","mean")
).round(1)

print("Table of car details:\n",carDetails)

Table of car details:
         cars  avg_mpg  avg_weight  avg_cylinders
origin                                          
europe    70     27.9      2423.3            4.2
japan     79     30.5      2221.2            4.1
usa      249     20.1      3361.9            6.2


Q8. size vs count

For each origin, show how many cars there are and how many have a
horsepower recorded.
Run this on a fresh copy of the data — if you filled the blanks in Q4, they are
gone.

In [81]:
carsCount=mpgDataSet.groupby("origin")['name'].size()
print("The number of cars from different origin:",carsCount)

The number of cars from different origin: origin
europe     70
japan      79
usa       249
Name: name, dtype: int64


In [82]:
horsepowerCount=mpgDataSet.groupby("origin")['horsepower'].count()
print("The number of horsepower:",horsepowerCount)

The number of horsepower: origin
europe     68
japan      79
usa       245
Name: horsepower, dtype: int64


Q9. Group by two columns

Average mpg for every combination of origin and cylinders . Then
reshape it with .unstack() so origins are rows and cylinder counts are
columns.

In [83]:
avgMsg=(mpg.groupby(['origin','cylinders'])['mpg'].mean()).round(2)
print("Average msg for different combination of origin and cylinders:",avgMsg)

unstackedTable=avgMsg.unstack()
print("The unstack table:\n",unstackedTable)

Average msg for different combination of origin and cylinders: origin  cylinders
europe  4            28.41
        5            27.37
        6            20.10
japan   3            20.55
        4            31.60
        6            23.88
usa     4            27.84
        6            19.66
        8            14.96
Name: mpg, dtype: float64
The unstack table:
 cylinders      3      4      5      6      8
origin                                      
europe       NaN  28.41  27.37  20.10    NaN
japan      20.55  31.60    NaN  23.88    NaN
usa          NaN  27.84    NaN  19.66  14.96


Q10. The confound

Someone shows you this:

Japanese cars average 30.5 mpg. American cars average 20.1.
Japanese engineering is simply better at fuel efficiency.

Both numbers are correct — you computed them yourself in Q7.
Is the conclusion correct? Investigate. Compare like with like, and state
clearly what the data does and does not support.

In [84]:
compareData=mpg.groupby("origin").agg(
    avgMpg=("mpg","mean"),
    avg_cylinders=("cylinders","mean"),
    avg_displacement=("displacement","mean"),
    avg_horsepower=("horsepower","mean"),
    avg_weight=("weight","mean"),
    avg_acceleration=("acceleration","mean"),
).round(2)

print("Comparing Data:\n",compareData)

Comparing Data:
         avgMpg  avg_cylinders  avg_displacement  avg_horsepower  avg_weight  \
origin                                                                        
europe   27.89           4.16            109.14           80.93     2423.30   
japan    30.45           4.10            102.71           79.84     2221.23   
usa      20.08           6.25            245.90          118.64     3361.93   

        avg_acceleration  
origin                    
europe             16.79  
japan              16.17  
usa                15.03  


Q11. idxmax

Which specific car has the best fuel economy in each origin? Show the car's
name and mpg, not just the number.

In [85]:
bestCarIndex=mpg.groupby("origin")['mpg'].idxmax()

bestCarDetails=mpg.loc[bestCarIndex,['name','mpg']]
print(bestCarDetails)

                     name   mpg
325  vw rabbit c (diesel)  44.3
322             mazda glc  46.6
344        plymouth champ  39.0


Q12. astype and to_csv

1. Convert origin to the category dtype and report the memory saved.
2. Save your Q7 summary table to origin_summary.csv — with the origin
column intact.
3. Read the file back and confirm nothing was lost.

In [ ]:
mpg['origin'].astype(mpg['name'].dtype)
carDetails=carDetails.reset_index(names="origin")

carDetails.to_csv("origin_summary.csv", index=False)


3


In [ ]:
summary=pd.read_csv("origin_summary.csv")
print(summary.shape[0])

# now also the origin number remains the same...hence no change

3


Q13. Sorting

Show the 10 heaviest cars, with name, weight and mpg. What do you notice
about their mpg?

In [89]:
mpg=pd.read_csv("mpg.csv")
mpgData=pd.read_csv("mpg.csv")

sortedData=mpg.sort_values(by="weight", ascending=False)
print(sortedData[['name','weight','mpg']].head(10))

                             name  weight   mpg
44            pontiac safari (sw)    5140  13.0
103              chevrolet impala    4997  11.0
42              dodge monaco (sw)    4955  12.0
90       mercury marquis brougham    4952  12.0
95       buick electra 225 custom    4951  12.0
104                  ford country    4906  12.0
43       ford country squire (sw)    4746  13.0
94   chrysler new yorker brougham    4735  13.0
28                       hi 1200d    4732   9.0
137      buick century luxus (sw)    4699  13.0


Q14. Compound filter

Find American cars that are 4-cylinder, from model year 1978 or later, and
manage more than 25 mpg. How many are there? Show the 5 most efficient.

In [92]:
americanCarDetails=mpg.loc[(mpg['origin']=="usa") & (mpg["model_year"]>=78) & (mpg['cylinders']==4) & (mpg['mpg']>25),:]

print("Number of filtered cars:",americanCarDetails.shape[0])
print("The american cars details with filter:\n",americanCarDetails.head())

Number of filtered cars: 37
The american cars details with filter:
       mpg  cylinders  displacement  horsepower  weight  acceleration  \
245  36.1          4          98.0        66.0    1800          14.4   
255  25.1          4         140.0        88.0    2720          15.4   
266  30.0          4          98.0        68.0    2155          16.5   
269  30.9          4         105.0        75.0    2230          14.5   
295  35.7          4          98.0        80.0    1915          14.4   

     model_year origin                         name  
245          78    usa                  ford fiesta  
255          78    usa          ford fairmont (man)  
266          78    usa           chevrolet chevette  
269          78    usa                   dodge omni  
295          79    usa  dodge colt hatchback custom  


Q15. String methods

The name column holds make and model together, like chevrolet chevelle
malibu .
1. Extract the brand — the first word — into a new brand column.
2. List every distinct brand.
3. Something is wrong with that list. Find the problem and fix it.

In [97]:
mpg['brand']=mpg['name'].str.strip().str.split()
mpg['brand']=mpg['brand'].str[0]
print(mpg['brand'])

0      chevrolet
1          buick
2       plymouth
3            amc
4           ford
         ...    
393         ford
394           vw
395        dodge
396         ford
397        chevy
Name: brand, Length: 398, dtype: object


In [98]:
print("Distinct Brand:\n", mpg['brand'].unique())

Distinct Brand:
 ['chevrolet' 'buick' 'plymouth' 'amc' 'ford' 'pontiac' 'dodge' 'toyota'
 'datsun' 'volkswagen' 'peugeot' 'audi' 'saab' 'bmw' 'chevy' 'hi'
 'mercury' 'opel' 'fiat' 'oldsmobile' 'chrysler' 'mazda' 'volvo' 'renault'
 'toyouta' 'maxda' 'honda' 'subaru' 'chevroelt' 'capri' 'vw'
 'mercedes-benz' 'cadillac' 'mercedes' 'vokswagen' 'triumph' 'nissan']


In [100]:
spellingCorrection={
    'chevy': 'chevrolet',
    'chevroelt': 'chevrolet',
    'toyouta': 'toyota',
    'maxda': 'mazda',
    'vokswagen': 'volkswagen',
    'vw': 'volkswagen',
    'mercedes': 'mercedes-benz',
    'hi': 'ihc'
}

mpg['brand']=mpg['brand'].replace(spellingCorrection)

print("Corrected Brand Spelling:\n",mpg['brand'].unique())

Corrected Brand Spelling:
 ['chevrolet' 'buick' 'plymouth' 'amc' 'ford' 'pontiac' 'dodge' 'toyota'
 'datsun' 'volkswagen' 'peugeot' 'audi' 'saab' 'bmw' 'ihc' 'mercury'
 'opel' 'fiat' 'oldsmobile' 'chrysler' 'mazda' 'volvo' 'renault' 'honda'
 'subaru' 'capri' 'mercedes-benz' 'cadillac' 'triumph' 'nissan']


Q16. Brand summary

Using your cleaned brand column: for every brand with at least 15 cars,
show the number of cars and the average mpg, sorted from most to least efficient.

In [102]:
carEfficiencyDetails=mpg.groupby("brand").agg(
    car_number=("mpg","count"),
    avg_mpg=("mpg","mean"),
).sort_values(by="avg_mpg").round(2)

print("The car filter brand:\n", carEfficiencyDetails.loc[carEfficiencyDetails["car_number"]>=15,:])

The car filter brand:
             car_number  avg_mpg
brand                          
amc                 28    18.25
buick               17    19.18
ford                51    19.69
pontiac             16    20.01
chevrolet           47    20.22
plymouth            31    21.70
dodge               28    22.06
toyota              26    28.17
datsun              23    31.11
volkswagen          22    31.84


Q17. Change over time

1. What was the average mpg of the 1970 fleet? The 1982 fleet?
2. Someone concludes: "Japanese cars got much more efficient over the
decade — that is why they took the market." Check it. Did Japanese cars
improve more than everyone else's?

In [106]:
avgMpg70=mpg.loc[mpg['model_year']==70,['mpg']].mean().round(2)
print("Average mpg of 1970 fleet:",avgMpg70)

avgMpg82=mpg.loc[mpg['model_year']==82,['mpg']].mean().round(2)
print("Average mpg of 1982 fleet:",avgMpg82)

Average mpg of 1970 fleet: mpg    17.69
dtype: float64
Average mpg of 1982 fleet: mpg    31.71
dtype: float64


In [ ]:
sortYearData=mpg.sort_values(by="model_year")
compareMpgMedian=sortYearData.groupby("origin")['mpg'].median().round(2)
compareMpgMean=sortYearData.groupby("origin")['mpg'].mean().round(2)

print("The comparison of MPG based on median:\n",compareMpgMedian)
print("The comparison of MPG based on mean:\n",compareMpgMean)

# As the median and mean of mpg for japan sorted based on year is more than any other => hence, it has evolved over time and the statement is true

The comparison of MPG based on median:
 origin
europe    26.5
japan     31.6
usa       18.5
Name: mpg, dtype: float64
The comparison of MPG based on mean:
 origin
europe    27.89
japan     30.45
usa       20.08
Name: mpg, dtype: float64


Q18. idxmin

For each cylinder count, which car has the worst fuel economy?

In [113]:
worstMpgIndex=mpg.groupby("cylinders")['mpg'].idxmin()
print("The car with worst MPG index:\n",mpg.loc[worstMpgIndex,"name"])

The car with worst MPG index:
 111          maxda rx3
76     volvo 145e (sw)
274          audi 5000
128     chevrolet nova
28            hi 1200d
Name: name, dtype: str


In [ ]:
compareTrendAccMpg=mpg[['mpg','acceleration']]
print("The mpg and acceleration relation: \n",compareTrendAccMpg.sort_values(by="mpg").head(10))

# By the general trend, we can see the lower the mpg, the higher the acceleration, hence it can be deduced that acceleration and mpg are inversely proportional

The mpg and acceleration relation: 
       mpg  acceleration
28    9.0          18.5
25   10.0          14.0
26   10.0          15.0
103  11.0          14.0
124  11.0          11.0
27   11.0          13.5
67   11.0          11.0
95   12.0          11.0
104  12.0          12.5
90   12.0          11.5
